### PR TITLE
Add support for Cohere and others.

### PR DESCRIPTION
**Pull request by austintlee**
### Description
Add support for Cohere and others.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ x] New functionality includes testing.
  - [ x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by austintlee** [comment]
Working on resolving a conflict caused by another PR, but this diff for this PR should be fairly small and clean.

**Comment by dhrubo-os** [comment]
Tests are failing.

**Comment by austintlee** [comment]
@dhrubo-os @HenryL27

File: plugin/src/test/java/org/opensearch/ml/rest/RestMLRAGSearchProcessorIT.java [hunk 1]
```diff
@@ -147,6 +148,44 @@ public class RestMLRAGSearchProcessorIT extends RestMLRemoteInferenceIT {
...
+        // + " \"post_process_function\": \"\\n String escape(def input) { \\n if (input.contains(\\\"\\\\\\\\\\\")) {\\n input =
+        // input.replace(\\\"\\\\\\\\\\\", \\\"\\\\\\\\\\\\\\\\\\\");\\n }\\n if (input.contains(\\\"\\\\\\\"\\\")) {\\n input =
+        // input.replace(\\\"\\\\\\\"\\\", \\\"\\\\\\\\\\\\\\\"\\\");\\n }\\n if (input.contains('\\r')) {\\n input = input =
+        // input.replace('\\r', '\\\\\\\\r');\\n }\\n if (input.contains(\\\"\\\\\\\\t\\\")) {\\n input = input.replace(\\\"\\\\\\\\t\\\",
+        // \\\"\\\\\\\\\\\\\\\\\\\\\\\\t\\\");\\n }\\n if (input.contains('\\n')) {\\n input = input.replace('\\n', '\\\\\\\\n');\\n }\\n if
+        // (input.contains('\\b')) {\\n input = input.replace('\\b', '\\\\\\\\b');\\n }\\n if (input.contains('\\f')) {\\n input =
+        // input.replace('\\f', '\\\\\\\\f');\\n }\\n return input;\\n }\\n def name = 'response';\\n def result = params.text;\\n def json
+        // = '{ \\\"name\\\": \\\"' + name + '\\\",' +\\n '\\\"dataAsMap\\\": { \\\"completion\\\": \\\"' + escape(result) +\\n '\\\"}}';\\n
+        // return json;\\n \\n \"\n"
```
Comment (dhrubo-os) [inline plugin/src/test/java/org/opensearch/ml/rest/RestMLRAGSearchProcessorIT.java:184]: do we need this?

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java [hunk 2]
```diff
@@ -105,7 +105,9 @@ protected Map<String, String> getInputParameters(ChatCompletionInput chatComplet
                 );
             inputParameters.put(CONNECTOR_INPUT_PARAMETER_MESSAGES, messages);
             // log.info("Messages to LLM: {}", messages);
-        } else if (chatCompletionInput.getModelProvider() == ModelProvider.BEDROCK) {
+        } else if (chatCompletionInput.getModelProvider() == ModelProvider.BEDROCK
+            || chatCompletionInput.getModelProvider() == ModelProvider.COHERE
+            || chatCompletionInput.getLlmResponseField() != null) {
```
Comment (dhrubo-os) [inline search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:110]: Seems like we need to update the else block accordingly.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java [hunk 3]
```diff
@@ -161,6 +175,19 @@ protected ChatCompletionOutput buildChatCompletionOutput(ModelProvider provider,
...
+            if (response != null) {
+                answers = List.of(response);
+            } else {
+                Map error = (Map) dataAsMap.get("error");
+                if (error != null) {
+                    errors = List.of((String) error.get("message"));
+                } else {
+                    errors = List.of("Unknown error or response.");
+                    log.error("{}", dataAsMap);
+                }
```
Comment (dhrubo-os) [inline search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:189]: seems like this block is completely similar except this line: `String response = (String) dataAsMap.get("text");` for bedrock it's `String response = (String) dataAsMap.get("completion");`.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java [hunk 4]
```diff
@@ -126,12 +128,24 @@ protected Map<String, String> getInputParameters(ChatCompletionInput chatComplet
...
+        if (llmResponseField != null) {
+            String response = (String) dataAsMap.get(llmResponseField);
+            if (response != null) {
+                answers = List.of(response);
+            } else {
+                Map error = (Map) dataAsMap.get("error");
+                if (error != null) {
+                    errors = List.of((String) error.get("message"));
+                } else {
+                    errors = List.of("Unknown error or response.");
```
Comment (dhrubo-os) [inline search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:145]: add test?

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/LlmIOUtil.java [hunk 5]
```diff
@@ -57,11 +60,19 @@ public static ChatCompletionInput createChatCompletionInput(
...
-        if (llmModel != null && llmModel.startsWith(BEDROCK_PROVIDER_PREFIX)) {
-            provider = Llm.ModelProvider.BEDROCK;
+        Llm.ModelProvider provider = null;
+        if (llmResponseField == null) {
+            provider = Llm.ModelProvider.OPENAI;
+            if (llmModel != null) {
+                if (llmModel.startsWith(BEDROCK_PROVIDER_PREFIX)) {
+                    provider = Llm.ModelProvider.BEDROCK;
+                } else if (llmModel.startsWith(COHERE_PROVIDER_PREFIX)) {
+                    provider = Llm.ModelProvider.COHERE;
```
Comment (dhrubo-os) [inline search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/LlmIOUtil.java:73]: add test?

Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/LlmIOUtil.java:73]: I do have tests for the new code.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 4 above)
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:145]: I do cover lines 142 - 143.  I ran the coverage report locally.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 3 above)
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:189]: Yeah, it's a bit inefficient.  I left it this way for now as I am not 100% how we'd converge on error handling across different models/providers.  I am focusing mostly on happy paths.  In the event of an error, this implementation does capture it (attempts to, at least).  But what I don't cover is all the different ways remote endpoints include errors in the responses.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 2 above)
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:110]: You mean we need to say something about llm_response_field?

File: plugin/src/test/java/org/opensearch/ml/rest/RestMLRAGSearchProcessorIT.java (hunk 1 above)
Comment (austintlee) [reply plugin/src/test/java/org/opensearch/ml/rest/RestMLRAGSearchProcessorIT.java:184]: No.  Will remove it.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 3 above)
Comment (dhrubo-os) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:189]: may be for now just put into a method expecting (text/completion) as parameter? At least that will reduce the lines :)

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 2 above)
Comment (dhrubo-os) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:110]: Yeah that's what I was wondering.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 4 above)
Comment (dhrubo-os) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:145]: interesting, why codecov is not reporting that?

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java [hunk 6]
```diff
@@ -119,19 +121,35 @@ protected Map<String, String> getInputParameters(ChatCompletionInput chatComplet
                         )
                 );
         } else {
-            throw new IllegalArgumentException("Unknown/unsupported model provider: " + chatCompletionInput.getModelProvider());
+            throw new IllegalArgumentException(
+                "Unknown/unsupported model provider: "
+                    + chatCompletionInput.getModelProvider()
+                    + ".  You must provide a valid model provider or llm_response_field."
```
Comment (HenryL27) [inline search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:127]: Not sure if its worth doing but maybe could add a list of the supported providers?

**Review by HenryL27 — APPROVED** [review]

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 6 above)
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:127]: I would just document it.  I don't want to update this string every time we add a new provider.

**Comment by austintlee** [comment]
@dhrubo-os

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 3 above)
Comment (dhrubo-os) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:193]: are we addressing this?

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 2 above)
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:110]: I updated it.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java [hunk 7]
```diff
@@ -119,19 +121,35 @@ protected Map<String, String> getInputParameters(ChatCompletionInput chatComplet
                         )
                 );
         } else {
-            throw new IllegalArgumentException("Unknown/unsupported model provider: " + chatCompletionInput.getModelProvider());
+            throw new IllegalArgumentException(
```
Comment (dhrubo-os) [inline search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:124]: add test?

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 4 above)
Comment (dhrubo-os) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:149]: oh ok, I was asking about the line no: 149

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/LlmIOUtil.java (hunk 5 above)
Comment (dhrubo-os) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/LlmIOUtil.java:73]: I don't see any test for cohere in the `LlmIOUtilTests`

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 4 above)
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java]: I added coverage for both cases in the new diff.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 3 above)
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java]: Yes, please check the latest diff.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 7 above)
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:125]: Sure.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java (hunk 4 above)
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java]: Addressed both cases in the new diff.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/LlmIOUtil.java (hunk 5 above)
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/LlmIOUtil.java:73]: Hmm.. This is covered in the IT tests I added, but I added UT to cover this.

**Review by dhrubo-os — APPROVED** [review]

**Review by HenryL27 — APPROVED** [review]

**Pull request merged**